c/extras/py_ftrl.cc

```diff
@@ -0,0 +1,413 @@
+  gs.add<&Ftrl::get_n_epochs, &Ftrl::set_n_epochs>("n_epochs", "Number of epochs to train for.\n");
+  gs.add<&Ftrl::get_inter, &Ftrl::set_inter>("inter", "If feature interactions to be used or not.\n");
+  gs.add<&Ftrl::get_hash_type, &Ftrl::set_hash_type>("hash_type", "Hashing method to use for strings.\n"
+    "`0` - std::hash;\n"
+    "`1` - Murmur2;\n"
+    "`2` - Murmur3.\n");
+  gs.add<&Ftrl::get_seed, &Ftrl::set_seed>("seed", "Seed to be used for Murmur hash functions.\n");
+
+  mm.add<&Ftrl::fit, args_fit>();
+  mm.add<&Ftrl::predict, args_predict>();
+  mm.add<&Ftrl::reset, args_reset>();
+}
+
+
+PKArgs Ftrl::Type::args_fit(1, 0, 0, false, false, {"frame"}, "fit",
+R"(fit(self, frame)
+--
+
+Train an FTRL model on a dataset.
+
+Parameters
+----------
+frame: Frame
+    Frame to be trained on, last column is treated as `target`.
+
+Returns
+----------
+    None
+)");
+
+
+void Ftrl::fit(const PKArgs& args) {
+  DataTable* dt_train = args[0].to_frame();
+  fm->fit(dt_train);
+}
+
+
+PKArgs Ftrl::Type::args_predict(1, 0, 0, false, false, {"frame"}, "predict",
+R"(predict(self, frame)
+--
+
+Make predictions for a dataset.
+
+Parameters
+----------
+frame: Frame
+    Frame of shape `(nrows, ncols)` to make predictions for. It must have one 
+    column less than the training dataset.
+
+Returns
+----------
+    A new `Frame` of shape `(nrows, 1)` with a prediction for each row.
+)");
+
+
+oobj Ftrl::predict(const PKArgs& args) {
+  if (fm->is_trained()) {
+    DataTable* dt_test = args[0].to_frame();
+    DataTable* dt_target = fm->predict(dt_test).release();
+    py::oobj df_target = py::oobj::from_new_reference(py::Frame::from_datatable(dt_target));
+    return df_target;
+  } else {
+    throw ValueError() << "Cannot make any predictions, because the model was not trained";
+  }
+}
+
+
+PKArgs Ftrl::Type::args_reset(0, 0, 0, false, false, {}, "reset",
+R"(reset(self)
+--
+
+Reset an FTRL model.
+
+Parameters
+----------
+    None
+
+Returns
+----------
+    None
+)");
+
+
+void Ftrl::reset(const PKArgs&) {
+  fm->init_model();
+}
+
+/*
+*  Getter and setter for the model datatable.
+*/
+oobj Ftrl::get_model() const {
+  if (fm->is_trained()) {
+    DataTable* dt_model = fm->get_model();
+    py::oobj df_model = py::oobj::from_new_reference(py::Frame::from_datatable(dt_model));
+    return df_model;
+  } else {
+    return py::None();
+  }
+}
+
+
+void Ftrl::set_model(robj model) {
+  DataTable* dt_model_in = model.to_frame();
+  const std::vector<std::string>& model_cols_in = dt_model_in->get_names();
+
+  if (dt_model_in->nrows != fm->get_d() || dt_model_in->ncols != 2) {
+    throw ValueError() << "FTRL model frame must have " << fm->get_d() << " rows,"
+                       << "and 2 columns, whereas your frame has " << dt_model_in->nrows
+                       << " rows and " << dt_model_in->ncols << " columns";
+  }
+
+  if (model_cols_in != FtrlModel::model_cols) {
```

Comment:
Right now the model frame that one can get with `model_frame = fm.model` has this kind of layout: two columns, names `z` and `n`, both `float64`. So there is no way to confuse these two sets of coefficients by looking at the frame content. Then there is a way to re-use the model with `fm.model = model_frame`, so I wanted to be as strict as possible and not load anything that just has two `float64` columns. Because the probability is higher that this frame is not a model frame, but something else. 

In principle, this could be replaced by some convention like `z` is the first column, and `n` is the second one, but then the probability is higher that some mess could be accepted as a model. Right now I just have the same requirements as to how the `model` frame should look like for both getter output and setter input.

Reply:
And there's nothing wrong with accepting "a mess" as a model. If a user wants to provide some weird initial values for the model, let them.
Only check the conditions that are strictly necessary. Like, verify that the coefficients do not have any NAs, and that n-values are all non-negative.

Reply:
Yes, we can accept a mess if needed. This check doesn't prevent it. What I wanted to do here is to have user be aware about it. So this won't happen just by an accident, because instead of a real model one just passed a dataset for predictions (since this one happens to have 2 `float64` columns). When you have proper names it is more likely that this frame is either already exported model, or user spent time by looking at the dataset.